### PR TITLE
Feature/SMC-75: Develop test suites for IP3 ERC-20 token part

### PR DESCRIPTION
**Pull request by marc-aurele-besner**
## Notable changes:
- Add more test for failure in Hardhat, event testing, and more
- Add more complex test in Foundry

### CODE REVIEW

File: test/IP3Token.EthlessBurn.test.js [hunk 1]
```diff
@@ -41,16 +41,52 @@ describe('IP3Token - Ethless Burn functions', function () {
...
+            ](owner.address, amountToBurn, feeToPay, nonce, signature);
+            await TestHelper.checkResult(input, IP3Token.address, user3, ethers, provider, 0);
+            expect(await IP3Token.balanceOf(owner.address)).to.equal(
+                ethers.BigNumber.from(originalBalance).sub(amountToBurn)
+            );
+            expect(await IP3Token.balanceOf(user3.address)).to.equal(ethers.BigNumber.from(feeToPay));
+        });
+    });
+
+    describe('IP3Token - Test expecting failure Ethless Burn', async function () {
```
Comment (kevinnguyen17) [inline test/IP3Token.EthlessBurn.test.js:55]: it is the failed case when the nonce was used only, try to add tests when different fee and amount

File: test/IP3Token.EthlessReserve.test.js [hunk 2]
```diff
@@ -59,10 +59,153 @@ describe('IP3Token - Ethless Reserve functions', function () {
...
+                owner.address,
+                nonce
+            );
+            await TestHelper.checkResult(inputReclaim, IP3Token.address, owner, ethers, provider, 0);
+            expect(await IP3Token.balanceOf(owner.address)).to.equal(ethers.BigNumber.from(originalBalance));
+            expect(await IP3Token.balanceOf(user1.address)).to.equal(ethers.BigNumber.from(0));
+        });
+    });
+
+    describe('IP3Token - Test expecting failure Ethless Reserve', async function () {
```
Comment (kevinnguyen17) [inline test/IP3Token.EthlessReserve.test.js:161]: it is the failed case when the nonce was used only, try to add tests when different expiration, fee and amount + after expiration

File: test/IP3Token.EthlessTransfer.test.js [hunk 3]
```diff
@@ -42,18 +42,54 @@ describe('IP3Token - Ethless Transfer functions', function () {
...
+            await TestHelper.checkResult(input, IP3Token.address, user3, ethers, provider, 0);
+            expect(await IP3Token.balanceOf(owner.address)).to.equal(
+                ethers.BigNumber.from(originalBalance).sub(amountToTransfer).sub(feeToPay)
+            );
+            expect(await IP3Token.balanceOf(user2.address)).to.equal(ethers.BigNumber.from(amountToTransfer));
+            expect(await IP3Token.balanceOf(user3.address)).to.equal(ethers.BigNumber.from(feeToPay));
+        });
+    });
+
+    describe('IP3Token - Test expecting failure Ethless Transfer', async function () {
```
Comment (kevinnguyen17) [inline test/IP3Token.EthlessTransfer.test.js:57]: it is the failed case when the nonce was used only, try to add tests when different fee and amount

**Review by kevinnguyen17 — CHANGES_REQUESTED** [review]
Add additional test cases as commented.

Create a common function to generate permit signature like this example https://github.com/gluwa/Mainnet-DAO/blob/dev/test/shared/signature.js

**Review by kevinnguyen17 — CHANGES_REQUESTED** [review]
The test suites are good. Besides, it is not necessary to test the failed txn due to reusing nonces.. in Foundry if you have done in Hardhat, instead you should compare the block numbers in the multiTxn scenario, we expect all these txns happen within the same block

**Review by ohyung1287 — APPROVED** [review]

**Review by kevinnguyen17 — APPROVED** [review]

**Pull request merged**